### PR TITLE
Fix hesse call limit

### PR DESCRIPTION
**Pull request by HDembinski**
This finally fixes https://github.com/scikit-hep/iminuit/issues/637

The following changes are implemented:

1) Store information that MnHesse reached call limit in MinimumError
- Add new states MnUnset, MnPosDef, and MnReachedCallLimit to MinimumError::Status
- Switch MinimumError status storage from bools to enum, since there is actually only one unique state at a time and toggling the bools correctly is error-prone; it also simplifies the implementation

2) Propagate information from MinimumError to FunctionMinimum via Add

I slightly simplified the implementation of VariableMetricBuilder, the variable `hessianComputed` was not needed, the information whether Hesse was run can be obtained from the MinimumState object.

Other changes (which could be broken into a separate PR):
- Removed trivial implementations of copy ctor and copy assignment in MnUserParameterState, the defaults should be used instead (rule of 0)

### CODE REVIEW

**Comment by mxxo** [comment]
@phsft-bot build

**Comment by HDembinski** [comment]
I cannot see why the builds failed...

**Comment by mxxo** [comment]
@phsft-bot build
I can't see any test failures on the dashboard, let's run it again and see.

File: math/minuit2/inc/Minuit2/MinimumError.h [hunk 1]
```diff
@@ -68,24 +64,23 @@ class MinimumError {
...
-   bool IsValid() const { return fPtr->fValid; }
-   bool IsPosDef() const { return fPtr->fPosDef; }
-   bool IsMadePosDef() const { return fPtr->fMadePosDef; }
-   bool HesseFailed() const { return fPtr->fHesseFailed; }
-   bool InvertFailed() const { return fPtr->fInvertFailed; }
-   bool IsAvailable() const { return fPtr->fAvailable; }
+   Status GetStatus() const { return fPtr->fStatus; }
+
+   bool IsValid() const { return IsAvailable() && (IsPosDef() || IsMadePosDef()); }
+   bool IsAccurate() const { return IsPosDef() && Dcovar() < 0.1; }
```
Comment (HDembinski) [inline math/minuit2/inc/Minuit2/MinimumError.h:70]: Before, this just checked Dcovar() < 0.1. Checking IsPosDef() is not necessary they way the ctors are set up, but this seems more clear and safer.

File: math/minuit2/inc/Minuit2/MnUserParameterState.h [hunk 2]
```diff
@@ -56,35 +56,6 @@ class MnUserParameterState {
...
-         fCovStatus = state.fCovStatus;
-         fFVal = state.fFVal;
-         fEDM = state.fEDM;
-         fNFcn = state.fNFcn;
-         fParameters = state.fParameters;
-         fCovariance = state.fCovariance;
-         fGlobalCC = state.fGlobalCC;
-         fIntParameters = state.fIntParameters;
-         fIntCovariance = state.fIntCovariance;
-      }
```
Comment (HDembinski) [inline math/minuit2/inc/Minuit2/MnUserParameterState.h:84]: This is a drive-by fix that could be split into a separate PR.

File: math/minuit2/src/MnUserParameterState.cxx [hunk 3]
```diff
@@ -112,7 +112,7 @@ MnUserParameterState::MnUserParameterState(const MnUserParameters &par, const Mn
 //
 //
 MnUserParameterState::MnUserParameterState(const MinimumState &st, double up, const MnUserTransformation &trafo)
-   : fValid(st.IsValid()), fCovarianceValid(false), fGCCValid(false), fCovStatus(-1), fFVal(st.Fval()), fEDM(st.Edm()),
+   : fValid(false), fCovarianceValid(false), fGCCValid(false), fCovStatus(-1), fFVal(st.Fval()), fEDM(st.Edm()),
```
Comment (HDembinski) [inline math/minuit2/src/MnUserParameterState.cxx:115]: fValid is later anyway reset

**Comment by HDembinski** [comment]
@lmoneta Thank you for accepting the other fix #8175, I now rebased this patch to the latest master and added some comments.

**Comment by HDembinski** [comment]
@lmoneta I cleaned up the patch a bit. The code works well, I am already using this patch in iminuit since May without issues. It would be great to get this fix also in ROOT.

**Comment by HDembinski** [comment]
@lmoneta Can we please merge this patch to ROOT? I have been using it for a while now in iminuit without issues. Without this being merged, I cannot move in iminuit from my patched version of ROOT back to the original ROOT master.

**Comment by lmoneta** [comment]
Sorry about the delay. I will check it on Monday and merge if it is ok.

**Comment by lmoneta** [comment]
@phsft-bot build

**Review by lmoneta — APPROVED** [review]
LGTM ! 
Thank you for the changes which fix the issue with the reached call limits for Hesse.
They also clear the handling of the different status of the MinimumError class. It looks also like everything is compatible as before.
I think we should update the Minuit2Minimizer::Hesse function to handle the new case of reached call limits, but this can be a separate PR. 
Apologies for the late review !
